o1vm/src/interpreters/riscv32im/interpreter.rs

```diff
@@ -1781,7 +1781,21 @@ pub fn interpret_itype<Env: InterpreterEnv>(env: &mut Env, instr: IInstruction)
             env.set_next_instruction_pointer(next_instruction_pointer + Env::constant(4u32));
         }
         IInstruction::ShiftRightArithmeticImmediate => {
-            unimplemented!("ShiftRightArithmeticImmediate")
+            // srai: x[rd] = x[rs1] >> shamt
+            let local_rs1 = env.read_register(&rs1);
+            let shamt = {
+                let pos = env.alloc_scratch();
+                unsafe { env.bitmask(&imm, 4, 0, pos) }
+            };
+            // parse shamt from imm as 20-24 of instruction and 0-4 wrt to imm
+            // sign extend shamt for arithmetic shift
+            let shamt = env.sign_extend(&shamt, 4);
+
+            let rd_scratch = env.alloc_scratch();
+            let local_rd = unsafe { env.shift_right_arithmetic(&local_rs1, &shamt, rd_scratch) };
```

Comment:
Nit: As in other places, I would like to keep the same shape:
```
let local_rd = {
  let pos = env.alloc_scratch();
  unsafe { env.shift_right_arithmetic(&local_rs1, &shamt, pos) }
};
```
Just to keep the scope of the position short.